src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq_vmr.c

```diff
@@ -3019,6 +3041,100 @@ static ssize_t program_vmr_store(struct device *dev,
 }
 static DEVICE_ATTR_WO(program_vmr);
 
+static ssize_t vmr_default_fpt_read(struct file *filp, struct kobject *kobj,
+        struct bin_attribute *attr, char *buf, loff_t off, size_t count)
+{
+        struct xocl_xgq_vmr *xgq =
+                dev_get_drvdata(container_of(kobj, struct device, kobj));
+        char *blob = NULL;
+        size_t size = 0;
+        ssize_t ret = 0;
+
+        /*
+         * refresh cached data only if off is 0,
+         * which indicates that there is a new request.
+         */
+        if (off == 0)
+                ret = xgq_refresh_default_fpt(xgq);
+
+        if (ret)
+                return ret;
+
+        mutex_lock(&xgq->xgq_lock);
+
+        blob = xgq->xgq_vmr_default_fpt;
+        size = xgq->xgq_vmr_default_fpt_size;
+
+        if (off >= size)
+                goto out;
+
+        if (off + count > size)
+                count = size - off;
+        memcpy(buf, blob + off, count);
+
+        ret = count;
+out:
+        mutex_unlock(&xgq->xgq_lock);
+        return ret;
+}
+
+static struct bin_attribute bin_attr_vmr_default_fpt = {
+        .attr = {
+                .name = "vmr_default_fpt",
+                .mode = 0444
+        },
+        .read = vmr_default_fpt_read,
+        .write = NULL,
+        .size = 0
+};
+
+static ssize_t vmr_backup_fpt_read(struct file *filp, struct kobject *kobj,
+        struct bin_attribute *attr, char *buf, loff_t off, size_t count)
+{
+        struct xocl_xgq_vmr *xgq =
+                dev_get_drvdata(container_of(kobj, struct device, kobj));
+        char *blob = NULL;
+        size_t size = 0;
+        ssize_t ret = 0;
+
+        /*
+         * refresh cached data only if off is 0,
+         * which indicates that there is a new request.
+         */
+        if (off == 0)
+                ret = xgq_refresh_backup_fpt(xgq);
+
+        if (ret)
+                return ret;
+
+        mutex_lock(&xgq->xgq_lock);
+
+        blob = xgq->xgq_vmr_backup_fpt;
+        size = xgq->xgq_vmr_backup_fpt_size;
+
+        if (off >= size)
+                goto out;
+
+        if (off + count > size)
+                count = size - off;
+        memcpy(buf, blob + off, count);
+
+        ret = count;
+out:
+        mutex_unlock(&xgq->xgq_lock);
+        return ret;
```

Comment:
worth creating a function for here and L3062-3078?

Reply:
there are indeed 3+ places dup code. could prapul fix this?